networkx/generators/tests/test_geometric.py

```diff
@@ -249,6 +249,18 @@ def test_navigable_small_world(self):
         gg = nx.grid_graph([5]).to_directed()
         assert nx.is_isomorphic(G, gg)
 
+    def test_invalid_diameter_value(self):
+        with pytest.raises(nx.NetworkXException, match=".*p must be >= 1"):
+            nx.navigable_small_world_graph(5, p=0, q=0, dim=1)
+
+    def test_invalid_long_range_connections_value(self):
+        with pytest.raises(nx.NetworkXException, match=".*q must be >= 0"):
+            nx.navigable_small_world_graph(5, p=1, q=-1, dim=1)
+
+    def test_invalid_exponent_for_decaying_probability_value(self):
+        with pytest.raises(nx.NetworkXException, match=".*r must be >= 0"):
+            nx.navigable_small_world_graph(5, p=1, q=0, r=-1, dim=1)
+
```

Comment:
+1 for this. I think we should also take this opportunity to test the result when `r` is between 0 and 1. AFAICT from skimming the [reference paper](https://dl.acm.org/doi/10.1145/335305.335325), this is indeed valid, but it'd be nice to have an explicit test (say for `r=0.5`)!

Reply:
@rossbar Added an explicit test case to test the scenario where `r` is a value between 0 and 1.

Reply:
Thanks @karthikchiru12 . Out of curiosity, where did this test case come from? Are we sure that the resulting graph is indeed a valid, navigable small world graph?

Reply:
I went ahead and dug a little deeper on this test case and ended up splitting it into two. The original test didn't probe the behavior of `r` because `q=0` means that no long-range edges were being added.

I went ahead and added a second test to probe the general behavior of `r` - basically larger `r` should result in fewer long-range edges. This is kind of tricky because it's probabilistic, but as long as `0 < q < number_of_nodes` this should hold in general.